types/validator_set.go

```diff
@@ -711,27 +710,97 @@ func (vals *ValidatorSet) VerifyFutureCommit(newSet *ValidatorSet, chainID strin
 		// }
 	}
 
-	if oldVotingPower <= oldVals.TotalVotingPower()*2/3 {
-		return errTooMuchChange{oldVotingPower, oldVals.TotalVotingPower()*2/3 + 1}
+	if got, needed := float32(oldVotingPower), float32(oldVals.TotalVotingPower()*2/3); got <= needed {
+		return ErrTooMuchChange{Got: got, Needed: needed}
+	}
+	return nil
+}
+
+// VerifyCommitTrusting verifies that trustLevel ([1/3, 1]) of the validator
+// set signed this commit.
+// NOTE the given validators do not necessarily correspond to the validator set
+// for this commit, but there may be some intersection.
+func (vals *ValidatorSet) VerifyCommitTrusting(chainID string, blockID BlockID,
+	height int64, commit *Commit, trustLevel float32) error {
+
+	if trustLevel > 1 || trustLevel < 1/3 {
+		panic(fmt.Sprintf("trustLevel must be within [1/3, 1], given %v", trustLevel))
+	}
+
+	if err := vals.verifyCommitBasic(commit, height, blockID); err != nil {
+		return err
+	}
+
+	talliedVotingPower := int64(0)
+	for idx, precommit := range commit.Precommits {
+		// skip absent and nil votes
+		// NOTE: do we want to check the validity of votes
+		// for nil?
+		if precommit == nil {
+			continue
+		}
+
+		// We don't know the validators that committed this block, so we have to
+		// check for each vote if its validator is already known.
+		_, val := vals.GetByAddress(precommit.ValidatorAddress)
+		if val != nil {
+			// Validate signature.
+			precommitSignBytes := commit.VoteSignBytes(chainID, idx)
+			if !val.PubKey.VerifyBytes(precommitSignBytes, precommit.Signature) {
+				return fmt.Errorf("Invalid commit -- invalid signature: %v", precommit)
+			}
+
+			// Good precommit!
+			if blockID.Equals(precommit.BlockID) {
+				talliedVotingPower += val.VotingPower
+			}
+			// else {
+			// It's OK that the BlockID doesn't match.  We include stray
+			// precommits to measure validator availability.
+			// }
+		}
+	}
+
+	if got, needed := float32(talliedVotingPower), float32(vals.TotalVotingPower())*trustLevel; got <= needed {
```

Comment:
Same concern about `float32` here.

Reply:
Yeah, but we can't multiply here because trustLevel is float32. We can probably add a new type 

```go
// Represents a fraction in terms of a numerator divided by a denominator. 
type Fraction struct {
	// The portion of the denominator in the faction, e.g. 2 in 2/3.
	Numerator in64
	// The value by which the numerator is divided, e.g. 3 in 2/3. Must be
	// positive.
	Denominator in64
}
```

Reply:
I don't think `trustLevel` should be `float32` either. A `Fraction` type sounds OK, or I think there is a fixed-precision `Decimal` in the Cosmos SDK that we could use.